File: app/src/main/java/com/simplepageindicator/PageViewIndicator.java
package com.simplepageindicator;

import android.content.Context;
import android.content.res.TypedArray;
import android.util.AttributeSet;
import android.util.Log;
import android.view.Gravity;
import android.view.ViewGroup;
import android.widget.LinearLayout;

/**
 * Created by renan on 12/11/15.
 */
public class PageViewIndicator extends LinearLayout {

    private int currentPage = 0;
    private int pageQuantity = 1;
    private int primaryColor = 0;
    private int secondaryColor = 0;
    private int indicadorWidth = 0;
    private int indicadorHeight = 0;
    private Context mContext;

    public PageViewIndicator(Context context, AttributeSet attrs) {
        super(context, attrs);
        this.mContext = context;
        setOrientation(LinearLayout.HORIZONTAL);
        setGravity(Gravity.CENTER_HORIZONTAL);
        TypedArray a = context.getTheme().obtainStyledAttributes(attrs,
                R.styleable.PageViewIndicator, 0, 0);
        try {
            //get attrs
            this.primaryColor = a.getInteger(R.styleable.PageViewIndicator_indicatorPrimaryColor, 0);
            this.secondaryColor = a.getInteger(R.styleable.PageViewIndicator_indicatorSecondaryColor, 0);
            this.indicadorWidth = a.getDimensionPixelSize(R.styleable.PageViewIndicator_indicatorWidth, 5);
            this.indicadorHeight = a.getDimensionPixelSize(R.styleable.PageViewIndicator_indicatorHeight, 5);
            this.pageQuantity = a.getInteger(R.styleable.PageViewIndicator_pageQuantity, 1);
            layoutWidthAdjust();
        } finally {
            a.recycle();
        }
        setPageQuantity(pageQuantity);
    }

    private void layoutWidthAdjust() {
        int width =  mContext.getResources().getDisplayMetrics().widthPixels;
        if((indicadorWidth * pageQuantity) >= width){
            this.indicadorWidth = width / pageQuantity ;
            this.indicadorHeight = width / pageQuantity ;
        }
    }

    public int getCurrentPage() {
        return currentPage;
    }

    public void nextPage(){
        clearPageIndicator(currentPage);
        if(currentPage < pageQuantity -1)
            currentPage++;
        else currentPage = 0;
        setCurrentPage(currentPage);
    }

    public void previousPage() {
        clearPageIndicator(currentPage);
        if(currentPage > 0)
            currentPage--;
        else currentPage = pageQuantity -1;
        setCurrentPage(currentPage);
    }

    public void setCurrentPage(int currentPage) {
        if(currentPage >= pageQuantity || currentPage < 0 ) return;
        ViewGroup mViewGroup = this;
        this.currentPage = currentPage;
        Indicator  indicator = (Indicator) mViewGroup.getChildAt(currentPage);
        indicator.setCircleColor(primaryColor);
    }

    public void clearPageIndicator(int currentPage){
        ViewGroup mViewGroup = this;
        Indicator indicator = (Indicator) mViewGroup.getChildAt(this.currentPage);
        indicator.setCircleColor(secondaryColor);
    }

    public void setPageQuantity(int quantity) {
        ViewGroup mViewGroup = this;
        this.pageQuantity = quantity;
        for (int i = 0; i < this.pageQuantity; i++) {
            Log.wtf("PageViewIndicator", "addView !!!");
            Indicator indicator = new Indicator(getContext());
            indicator.setLayoutParams(new LayoutParams(indicadorWidth, indicadorHeight));
            if (currentPage == i) {
                indicator.setCircleColor(primaryColor);
            } else {
                indicator.setCircleColor(secondaryColor);
            }
            LinearLayout.LayoutParams params = new LinearLayout.LayoutParams(ViewGroup.LayoutParams.WRAP_CONTENT, ViewGroup.LayoutParams.WRAP_CONTENT);
            mViewGroup.addView(indicator, i);
        }
    }

}

File: simplepageindicator/src/main/java/com/github/simplepageindicator/PageViewIndicator.kt
package com.github.simplepageindicator

import android.content.Context
import android.util.AttributeSet
import android.view.Gravity
import android.widget.LinearLayout
import com.simplepageindicator.R

class PageViewIndicator(
    private val mContext: Context,
    attrs: AttributeSet?
) : LinearLayout(mContext, attrs) {
    private var currentPage = 0
    private var pageNumber = 1
    private var primaryColor = 0
    private var secondaryColor = 0
    private var indicatorWidth = 0
    private var indicatorHeight = 0
    private var overflowMode = false

    init {
        orientation = HORIZONTAL
        gravity = Gravity.CENTER_HORIZONTAL
        val typedArray = mContext.theme.obtainStyledAttributes(
            attrs,
            R.styleable.PageViewIndicator, 0, 0
        )
        primaryColor = typedArray.getInteger(R.styleable.PageViewIndicator_indicatorPrimaryColor, 0)
        secondaryColor = typedArray.getInteger(R.styleable.PageViewIndicator_indicatorSecondaryColor, 0)
        indicatorWidth = typedArray.getDimensionPixelSize(R.styleable.PageViewIndicator_indicatorWidth, 5)
        indicatorHeight = typedArray.getDimensionPixelSize(R.styleable.PageViewIndicator_indicatorHeight, 5)
        pageNumber = typedArray.getInteger(R.styleable.PageViewIndicator_pageNumber, 1)
        overflowMode = typedArray.getBoolean(R.styleable.PageViewIndicator_overflowMode, false)

        typedArray.recycle()
        layoutWidthAdjust()
        setPageNumber(pageNumber)
    }

    private fun layoutWidthAdjust() {
        val width = mContext.resources.displayMetrics.widthPixels
        if (indicatorWidth * pageNumber >= width) {
            indicatorWidth = width / pageNumber
            indicatorHeight = width / pageNumber
        }
    }

    fun setPageNumber(number: Int) {
        removeAllViews()
        pageNumber = number
        layoutWidthAdjust()
        for (i in 0 until pageNumber) {
            val indicator = Indicator(context).apply {
                layoutParams = LayoutParams(indicatorWidth, indicatorHeight)
                setCircleColor(secondaryColor)
            }
            addView(indicator, i)
        }
        setCurrentPage(getCurrentPage())
    }

    fun setCurrentPage(pageIndex: Int) {
        clearPageIndicator()
        if (currentPage >= pageNumber || currentPage < 0) throw  Exception("Invalid page number")
        currentPage = pageIndex
        val indicator = getChildAt(currentPage) as Indicator
        indicator.setCircleColor(primaryColor)
    }

    private fun clearPageIndicator() {
        val indicator = getChildAt(currentPage) as Indicator
        indicator.setCircleColor(secondaryColor)
    }

    fun nextPage() {
        clearPageIndicator()
        if (currentPage < pageNumber - 1) {
            currentPage++
        } else if (overflowMode) {
            currentPage = 0
        }
        setCurrentPage(currentPage)
    }

    fun previousPage() {
        clearPageIndicator()
        if (currentPage > 0) {
            currentPage--
        } else if (overflowMode) {
            currentPage = pageNumber - 1
        }
        setCurrentPage(currentPage)
    }

    fun getCurrentPage(): Int {
        return currentPage
    }

}